### PR TITLE
classloader update

### PR DESCRIPTION
**Pull request by sanjaykumarcg**
Issue : class loader

i done the changes in pom.xml  replace SNAPSHOT with beta2 .
i have maked changes in ParsingUtils.java   and parser.parseQualifiedName(reader, file) replace with parser.parseType(reader, file) and return with getQualifiedName() . 

its returns only simple entity name  like EmployeeEntity ,so its not able to load class ;

### CODE REVIEW

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java [hunk 1]
```diff
@@ -385,7 +384,7 @@ private void setOutputRootPath(File inputProject) {
                         : ((TemplateTo) matching.get(index)).getId();
                     logger.info("(" + selectedArtifactNumber + ") " + artifactDescription);
                 } catch (IndexOutOfBoundsException e) {
-                    logger.error("The " + artifactType + " number you have specified is out of bounds!");
+                    logger.error("The " + artifactType + " number you have specified is out of bounds ,So please provide the number(s) of increments that you want to generate separated by comma.!");
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java]:
```Java
So please provide the number(s) of increments that you want to generate separated by comma
```
That message doesn't make sense to me. This exception is thrown when the user selects an increment number that doesn't exist (for instance `459`, we only have 15 increments at most). Therefore the message you have added is not needed

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java [hunk 2]
```diff
@@ -421,10 +420,14 @@ private void setOutputRootPath(File inputProject) {
                     logger.info("Exact match found: " + artifactDescription + ".");
                     userSelection.add(possibleArtifacts.get(0));
                     return userSelection;
+                }else if(possibleArtifacts.size()<1) {
+                	logger. info("No increment with that name has been found, Please provide correct increment name and try again ! Thank you");
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java]: Great, this is what I wanted :)

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java [hunk 3]
```diff
@@ -326,13 +326,12 @@ public void generate(File inputFile, File inputProject, List<? extends Generable
      * @param inputProject
      *            project where the code will be generated to
      */
-    private void setOutputRootPath(File inputProject) {
-        logger.info("As you did not specify where the code will be generated, we will use the project of your current"
-            + " input file.");
-        logger.debug("Generating to: " + inputProject.getAbsolutePath());
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java]: Please don't remove these two messages, they are important to have

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java [hunk 4]
```diff
@@ -421,10 +420,14 @@ private void setOutputRootPath(File inputProject) {
...
+                	logger. info("No increment with that name has been found, Please provide correct increment name and try again ! Thank you");
+                       
+                        System.exit(1);
                 }
-
-                logger.info("Please enter the number(s) of " + artifactType
-                    + "(s) that you want to generate separated by comma.");
+                
+                /*logger.info("Please enter the number(s) of " + artifactType
+                    + "(s) that you want to generate separated by comma.");*/
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java]: Can you explain me why have you removed this message?

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java [hunk 5]
```diff
@@ -364,25 +365,25 @@ public File getTemplatesJar(boolean isSource) {
...
      *            list containing elements to be retained in this list
-     * @param listToIntersect
+     * @param list2
      *            second list to be used for the intersection
      * @return <tt>resultant list</tt> containing increments that are in both lists
      */
-    public static List<IncrementTo> retainAllIncrements(List<IncrementTo> currentList,
-        List<IncrementTo> listToIntersect) {
+    public static List<IncrementTo> retainAllIncrements(List<IncrementTo> list,
+        List<IncrementTo> list2) {
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java]: Why have you changed the names of the variables? They had good names previously

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java [hunk 6]
```diff
@@ -57,19 +57,49 @@
...
 
-        JavaContext context = JavaSourceProviderUsingMaven.createFromLocalMavenProject(inputProject);
-
+        JavaContext context = JavaSourceProviderUsingMaven.createFromLocalMavenProject(inputProject,true);
+        String fqn = ParsingUtils.getFQN(inputFile);
         String qualifiedName = ParsingUtils.getQualifiedName(inputFile, context);
 
         try {
-            context.getClassLoader().loadClass(qualifiedName);
+           // context.getClassLoader().loadClass(qualifiedName);
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java]: Please remove this comment

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java [hunk 7]
```diff
@@ -57,19 +57,49 @@
...
+     * @return
+     */
+    private static String getFQN(File inputFile) {
+        String simpleName = inputFile.getName().replaceAll("\\.(?i)java", "");
+        String packageName = getPackageName(inputFile.getParentFile(), "");
+
+         return packageName + "." + simpleName;
+    }
+
+     private static String getPackageName(File folder, String packageName) {
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java]: Can you add Javadoc? This method is traversing parent folders until it reaches `java` folder in order to get the FQN

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java (hunk 4 above)
Comment (sanjaykumarcg) [reply cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java]: Actually this message is comes under If String representation is given,so dont need this message here .

Comment (jdiazgon) [reply cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java]: You are right that it comes under String representation, but please don't remove this message because in this part of the code (line 415):

```Java
if (possibleArtifacts.size() > 1) {
    printFoundArtifacts(possibleArtifacts, isIncrements, artifactType);
} 
```
We are printing all the found artifacts by number, like this:

```
(0) All
(1) TO's
(2) CRUD Rest services
...
```
Therefore that message makes sense, because the user needs to use numbers to select the option.

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java [hunk 8]
```diff
@@ -384,7 +386,7 @@ private void setOutputRootPath(File inputProject) {
                         : ((TemplateTo) matching.get(index)).getId();
                     logger.info("(" + selectedArtifactNumber + ") " + artifactDescription);
                 } catch (IndexOutOfBoundsException e) {
-                    logger.error("The " + artifactType + " number you have specified is out of bounds ,So please provide the number(s) of increments that you want to generate separated by comma.!");
+                    logger.error("The " + artifactType + " number you have specified is out of bound !");
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/GenerateCommand.java]: It's `out of bounds` you need the final `s` ;)

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java [hunk 9]
```diff
@@ -185,7 +185,7 @@ public CobiGen initializeCobiGen() {
         CobiGen cg = null;
         if (templatesJar != null) {
             try {
-                registerPlugins();
+            	registerPlugins();
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java]: Here you have added some extra tabs that we don't need, please remove those whitespaces

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java [hunk 10]
```diff
@@ -59,30 +59,34 @@ public static JavaContext getJavaContext(File inputFile, File inputProject) {
...
     }
-
-     private static String getPackageName(File folder, String packageName) {
+/**
+ *  This method traverse the folder in reverse order from child to parent
+ *  @param folder parent input file
+ *  @param package name 
+ *  @return package name
+ * */
+     @SuppressWarnings("javadoc")
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java]: The javadoc is perfect, but I don't understand why you have added `@SuppressWarnings("javadoc")`. Please remove it

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java [hunk 11]
```diff
@@ -8,7 +8,7 @@
  * This class defines the main CobiGen command
  */
 @Command(description = MessagesConstants.WELCOME_MESSAGE, name = "cobigen", aliases = { "cg" },
-    mixinStandardHelpOptions = true, version = "CobiGen CLI 1.0", subcommands = { GenerateCommand.class })
+    mixinStandardHelpOptions = true, version = {"CobiGen CLI 1.0","Java plug-in 2.1.0","OpenAPI plug-in	2.3.0","HTML plug-in 2.0.1"}, subcommands = { GenerateCommand.class })
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java]: This is related to issue #956, right? Actually that is not enough, please read again the issue.

What we need to do is read the artificial pom, in order to parse the different versions of the plug-ins and add them there. If you have any question regarding this, contact me.

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java [hunk 12]
```diff
@@ -178,29 +179,29 @@ public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOExce
      * Registers CobiGen plug-ins and instantiates CobiGen
      * @return object of CobiGen
      */
-    public CobiGen initializeCobiGen() {
-        getTemplates();
+	public CobiGen initializeCobiGen() {
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java]: Hmm these tab characters doesn't look good. Can you revert these changes?

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java [hunk 13]
```diff
@@ -249,7 +251,7 @@ public void registerPlugins() {
             logger.error("Not able to convert current location of the CLI to URI. Most probably this is a bug", e);
         }
 
-    }
+    };
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java]: This `;` is not actually needed.

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java [hunk 14]
```diff
@@ -364,9 +366,9 @@ public File getTemplatesJar(boolean isSource) {
      * specified collection (optional operation). In other words, the resultant list removes from this list
      * all of its elements that are not contained in the specified collection.
      *
-     * @param currentList
+     * @param list
      *            list containing elements to be retained in this list
-     * @param listToIntersect
+     * @param list2
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java]: Please change the JavaDoc back to `currentList` and `listToIntersect`

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java [hunk 15]
```diff
@@ -48,19 +57,53 @@
...
+        String packageName = getPackageName(inputFile.getParentFile(), "");
+
+         return packageName + "." + simpleName;
+    }
+/**
+ *  This method traverse the folder in reverse order from child to parent
+ *  @param folder parent input file
+ *  @param packageName 
+ *  @return package name
+ * */
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/ParsingUtils.java]: Please format this correctly, the JavaDoc needs to be aligned with the method declaration.

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java [hunk 16]
```diff
@@ -1,20 +1,74 @@
...
+			MavenXpp3Reader reader = new MavenXpp3Reader();
+			Model model;
+			if ((new File("pom.xml")).exists())
+				model = reader.read(new FileReader("pom.xml"));
+			else
+				model = reader.read(
+						new InputStreamReader(CobiGenCommand.class.getResourceAsStream("/src/main/resources/pom.xml")));
+			List<Model> versionname = new ArrayList<Model>();
+			versionname.add(model);
+			for (int i = 0; i < versionname.size(); i++)
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java]: I think you don't need this `for` loop.

So you have created a List called `versionname` where you add the model: `versionname.add(model)`

Why don't you directly use `model` itself in order to get the dependencies?

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java [hunk 17]
```diff
@@ -1,20 +1,74 @@
...
+				for (int j = 0; j < versionname.get(i).getDependencies().size(); j++) {
+					System.out.println("name:= "+versionname.get(i).getName()+"  version"+versionname.get(i).getVersion());
+					System.out.println("name:= " + versionname.get(i).getDependencies().get(j).getArtifactId()
+							+ "  version= " + versionname.get(i).getDependencies().get(j).getVersion());
+				}
+			String s[] = new String[10];
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java]: It seems that you need to add to that array all the different artifacts and versions, like:

```Java
s[0] = ""name:= " + versionname.get(i).getDependencies().get(j).getArtifactId()
							+ "  version= " + versionname.get(i).getDependencies().get(j).getVersion())
```

Therefore, you don't need those `System.out.println` as Picocli will use that array in order to print the versions

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java [hunk 18]
```diff
@@ -1,20 +1,74 @@
...
+		@Override
+		public String[] getVersion() throws Exception {
+
+			MavenXpp3Reader reader = new MavenXpp3Reader();
+			Model model;
+			if ((new File("pom.xml")).exists())
+				model = reader.read(new FileReader("pom.xml"));
+			else
+				model = reader.read(
+						new InputStreamReader(CobiGenCommand.class.getResourceAsStream("/src/main/resources/pom.xml")));
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java]: Actually, we need to read the pom.xml that is present on the cobigen-cli folder. On the first execution of the CLI, we extract that pom.xml next to the `cobigen.jar` file. You should read that pom, because in the future, we will be able to change versions directly to that pom.

If you take a look [here](https://github.com/devonfw/tools-cobigen/blob/master/cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java#L230), you will see how to get the real path of the pom.xml.

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java [hunk 19]
```diff
@@ -1,20 +1,74 @@
...
+		// Nothing to do here, this is the master command
+	}
+
+	static class PropertiesVersionProvider implements IVersionProvider {
+		@Override
+		public String[] getVersion() throws Exception {
+
+			MavenXpp3Reader reader = new MavenXpp3Reader();
+			Model model;
+			if ((new File("pom.xml")).exists())
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java]: Is this always returning `false`? If so, please look on the comment below.

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java [hunk 20]
```diff
@@ -375,20 +376,21 @@ public File getTemplatesJar(boolean isSource) {
...
+				String intersectIncrementDesc = ((IncrementTo) intersectIncrement).getDescription().trim()
+						.toLowerCase();
+
+				if (currentIncrementDesc.equals(intersectIncrementDesc)) {
+					resultantList.add((IncrementTo) currentIncrement);
+					break;
+				}
+			}
+		}
+		return resultantList;
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/utils/CobiGenUtils.java]: Can you revert these changes? I think they are not needed.

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java [hunk 21]
```diff
@@ -46,30 +50,58 @@ public void run() {
...
+			List<String> versionProvider = new ArrayList<String>();
 			MavenXpp3Reader reader = new MavenXpp3Reader();
-			Model model;
-			if ((new File("pom.xml")).exists())
-				model = reader.read(new FileReader("pom.xml"));
-			else
-				model = reader.read(
-						new InputStreamReader(CobiGenCommand.class.getResourceAsStream("/src/main/resources/pom.xml")));
+			Model model = null;
+			File locationCLI = new File(CobiGenUtils.class.getProtectionDomain().getCodeSource().getLocation().toURI());
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java]: Instead of `CobiGenUtils` can you use `CobiGenCommand`? It is the class that we currently are :)

File: cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java [hunk 22]
```diff
@@ -46,30 +50,58 @@ public void run() {
...
 	}
 
+	/**
+	 * Extracts an artificial POM which defines all the CobiGen plug-ins that are
+	 * needed
+	 * 
+	 * @param rootCLIPath path where the artificial POM will be extracted to
+	 * @return the extracted POM file
+	 */
+	private static File extractArtificialPom(Path rootCLIPath) {
```
Comment (jdiazgon) [inline cobigen-cli/src/main/java/com/devonfw/cobigen/cli/commands/CobiGenCommand.java]: Here you are duplicating this method from CobiGenUtils. Please call directly the method from there, make it `public` if needed
